pkg/font/font.go

```diff
@@ -0,0 +1,259 @@
+			largestBearingY = info.bearingY
+		}
+	}
+	return largestBearingY
+}
+
+// MapString turns each character in the string into a pair of
+// (x,y,s,t)-vertex triangles using glyph information from a
+// pre-loaded font. The vertex info is returned as []float32
+func MapString(str string, font Info, pos sdl.Point, align ui.Align) []float32 {
+	return MapStringWithBearing(str, font, GetMaxVerticalBearing(str, font), pos, align)
+}
+
+// MapStringWithBearing turns each character in the string into
+// a pair of (x,y,s,t)-vertex triangles using glyph information
+// from a pre-loaded font and the font information, with the
+// vertical bearing provided. The vertex info is returned as []float32
+func MapStringWithBearing(str string, font Info, maxBearingY float32, pos sdl.Point, align ui.Align) []float32 {
+	// 2 triangles per rune, 3 vertices per triangle, 4 float32's per vertex (x,y,s,t)
+	buffer := make([]float32, 0, len(str)*24)
+	// get glyph information for alignment
+	var strWidth float32
+	for _, r := range str {
+		info := font.runeMap[r-minASCII]
+		strWidth += info.advance
+	}
+	// adjust strWidth if last rune's width + bearingX > advance
+	lastInfo := font.runeMap[str[len(str)-1]-minASCII]
+	if float32(lastInfo.width)+lastInfo.bearingX > lastInfo.advance {
+		strWidth += (float32(lastInfo.width) + lastInfo.bearingX - lastInfo.advance)
+	}
+
+	w2 := float64(strWidth) / 2.0
+	h2 := float64(font.textHeight) / 2.0
+	offx := int32(-w2 - float64(align.H)*w2)
+	offy := int32(-h2 - float64(align.V)*h2)
+	// offset origin to account for alignment
+	origin := pointF32{float32(pos.X + offx), float32(pos.Y+offy) + maxBearingY}
+	for _, r := range str {
+		info := font.runeMap[r-minASCII]
+
+		// calculate x,y position coordinates - use bottom left as (0,0); shader converts for you
+		posTL := pointF32{origin.x + info.bearingX, origin.y + (float32(info.height) - info.bearingY)}
+		posTR := pointF32{posTL.x + float32(info.width), posTL.y}
+		posBL := pointF32{posTL.x, origin.y - info.bearingY}
+		posBR := pointF32{posTR.x, posBL.y}
+		// calculate s,t texture coordinates - use top left as (0,0); shader converts for you
+		texTL := pointF32{0, float32(info.row)}
+		texTR := pointF32{float32(info.width), texTL.y}
+		texBL := pointF32{texTL.x, texTL.y + float32(info.height)}
+		texBR := pointF32{texTR.x, texBL.y}
+		// create 2 triangles
+		triangles := []float32{
+			posBL.x, posBL.y, texBL.x, texBL.y, // bottom-left
+			posTL.x, posTL.y, texTL.x, texTL.y, // top-left
+			posTR.x, posTR.y, texTR.x, texTR.y, // top-right
+
+			posBL.x, posBL.y, texBL.x, texBL.y, // bottom-left
+			posTR.x, posTR.y, texTR.x, texTR.y, // top-right
+			posBR.x, posBR.y, texBR.x, texBR.y, // bottom-right
+		}
+		buffer = append(buffer, triangles...)
+
+		origin.x += info.advance
+	}
+
+	return buffer
+}
+
+type fontKey struct {
+	fontName string
+	fontSize int32
+}
+
+type Info struct {
+	TextureID  uint32     // OpenGL texture ID of cached glyph data
```

Comment:
It seems like this should not be capitalized

Reply:
Done.